sdk/python/kfp/dsl/databricks/_run_op.py

```diff
@@ -0,0 +1,264 @@
+        spec: {} = None,
+        existing_cluster_id: str = None,
+        new_cluster: {} = None,
+        libraries: {} = None,
+        spark_jar_task: {} = None,
+        spark_python_task: {} = None,
+        spark_submit_task: {} = None,
+        notebook_task: {} = None,
+        timeout_seconds: int = None):
+        """Create a new instance of SubmitRunOp.
+
+        Args:
+
+            name: the name of the op. It does not have to be unique within a pipeline
+                because the pipeline will generates a unique new name in case of conflicts.
+            run_name: an optional name for the Spark Run. The default value is Untitled.
+            spec: full specification of the Spark run to submit. If you provide this, 
+                all other args except for name and run_name will be ignored.
+            existing_cluster_id: the ID of an existing cluster that will be used for all runs of this job.
+            new_cluster: A description of a cluster that will be created for each run
+            libraries: An optional list of libraries to be installed on the cluster that will execute the job.
+            spark_jar_task: Indicates that this job should run a JAR.
+            spark_python_task: Indicates that this job should run a Python file.
+            spark_submit_task: Indicates that this job should run spark submit script.
+            notebook_task: Indicates that this job should run a notebook.
+            timeout_seconds: An optional timeout applied to each run of this job. 
+                The default behavior is to have no timeout.
+                
+        Raises:
+
+            ValueError: if not inside a pipeline
+                        if the name is an invalid string
+                        if no cluster id or description is provided
+                        if both cluster id and description are provided
+                        if no spark jar task or spark python task or spark submit task definition is provided
+                        if two or more task definitions are provided
+        """
+
+        if not spec:
+            spec = {}
+
+            if not existing_cluster_id and not new_cluster:
+                raise ValueError("You need to provide a existing_cluster_id or a new_cluster.")
+            elif existing_cluster_id and new_cluster:
+                raise ValueError("You cannot provide both a existing_cluster_id and a new_cluster.")
+            elif not existing_cluster_id:
+                spec["new_cluster"] = new_cluster
+            else:
+                spec["existing_cluster_id"] = existing_cluster_id
+
+            if not spark_jar_task and not spark_python_task and not spark_submit_task and not notebook_task:
+                raise ValueError("You need to provide a spark_jar_task or a spark_python_task or a spark_submit_task or a notebook_task.")
+            elif spark_jar_task and not spark_python_task and not spark_submit_task and not notebook_task:
```

Comment:
this logic looks a bit strange, personally I would say that the DSL should probably just map the spark tasks and notebook tasks and then let the operator understand if the combination is valid

I fear if this logic is here then it means we have duplication of business concerns between the operator and the DSL

Reply:
These business rules are documented in Databricks API. I agree we may be duplicating business rules. But if we fail the checks in the DSL we wouldn't be creating invalid k8s CRDs and making the operator to work for nothing. I was actually thinking about the possibility of doing these checks when we get the whole spec, too. Although implementing and maintaining the Databricks DSL would be much easier if we just don't care what we get from the user and pass it through to the operator.

Reply:
I've been doing further tests, and apparently Databricks API won't return any error if the different parts of the spec don't follow the rules specified in the docs. Databricks will try to execute the run anyway. So I guess it's ok to remove these checks from the DSL if the API itself is not enforcing them.

Reply:
I removed the additional checks